compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/PropertyReferenceLowering.kt

```diff
@@ -201,223 +184,253 @@ internal class PropertyReferenceLowering(val context: JvmBackendContext) : Class
+                    putValueArgument(0, calculateOwner(expression.propertyContainer, this@PropertyReferenceLowering.context))
+                    putValueArgument(1, irString(expression.referencedName.asString()))
+                    putValueArgument(2, computeSignatureString(expression))
+                })
             }
+        }
+    }
 
-            // Create an instance of KProperty that overrides the get() and set() methods to directly call getX() and setX() on the object.
-            // This is (relatively) fast, but space-inefficient. Also, the instances can store bound receivers in their fields. Example:
-            //
-            //    class C$property$0 : PropertyReference0Impl {
-            //        constructor(boundReceiver: C) : super(boundReceiver, C::class.java, "property", "getProperty()LType;", 0)
-            //        override fun get(): T = receiver.property
-            //        override fun set(value: T) { receiver.property = value }
-            //    }
-            //
-            // and then `C()::property` -> `C$property$0(C())`.
-            //
-            private fun createSpecializedKProperty(expression: IrCallableReference<*>): IrExpression {
-                val referenceClass = createKPropertySubclass(expression)
-                return context.createIrBuilder(
-                        currentScope?.scope?.scopeOwnerSymbol ?: irClass.symbol, expression.startOffset, expression.endOffset
-                    )
-                    .irBlock {
-                        // TODO: Move this to the enclosing class, right now the parent field is wrong!
-                        +referenceClass
-                        +irCall(referenceClass.constructors.single()).apply {
-                            var index = 0
-                            expression.dispatchReceiver?.let { putValueArgument(index++, it) }
-                            expression.extensionReceiver?.let { putValueArgument(index++, it) }
-                        }
-                    }
+    // Create an instance of KProperty that overrides the get() and set() methods to directly call getX() and setX() on the object.
+    // This is (relatively) fast, but space-inefficient. Also, the instances can store bound receivers in their fields. Example:
+    //
+    //    class C$property$0 : PropertyReference0Impl {
+    //        constructor(boundReceiver: C) : super(boundReceiver, C::class.java, "property", "getProperty()LType;", 0)
+    //        override fun get(): T = receiver.property
+    //        override fun set(value: T) { receiver.property = value }
+    //    }
+    //
+    // and then `C()::property` -> `C$property$0(C())`.
+    //
+    private fun createSpecializedKProperty(expression: IrCallableReference<*>): IrExpression {
+        // We do not reuse classes for non-reflective property references because they would not have
+        // a valid enclosing method if the same property is referenced at many points.
```

Comment:
Strictly speaking, we don't have (and JLS/JVMS don't impose) any such requirements on EnclosingMethod, so we can start reusing classes for references at some point ([KT-15690](https://youtrack.jetbrains.com/issue/KT-15690)). But I'd prefer to do this after JVM IR is stabilized.